qcengine/cli.py

```diff
@@ -4,29 +4,147 @@
+    run_procedure.add_argument('procedure', type=str, help="The procedure to run.")
+    run_procedure.add_argument('data', type=str, help="Data describing the task to run. "
+                                                      "One of: (i) A JSON blob. "
+                                                      "(ii) A file name. "
+                                                      "(iii) '-', indicating data will be read from STDIN.")
+
+    args = vars(parser.parse_args())
+    if args["command"] is None:
+        parser.print_help(sys.stderr)
+        exit(1)
+
+    # Print usage and exit if no info options are provided
+    if args["command"] == "info" and not any([v for v in args.values() if v != "info"]):
+        info.print_help()
+        exit(1)
+
+    return args
+
+
+def info_cli(args):
+
+    def info_version():
+        import qcelemental
+        print(">>> Version information")
+        print(f"QCEngine version:    {__version__}")
+        print(f"QCElemental version: {qcelemental.__version__}")
+        print()
+
+    def info_programs():
+        print(">>> Program information")
+        all_progs = list_all_programs()
+        avail_progs = list_available_programs()
+        print("Available programs:")
+        for prog_name in sorted(avail_progs):
+            version = get_program(prog_name).get_version()
+            if version is None:
+                version = "???"
+            print(f"{prog_name} v{version}")
+
+        print()
+        print("Other supported programs:")
+        print(" ".join(sorted(all_progs - avail_progs)))
+        print()
+
+    def info_procedures():
+        print(">>> Procedure information")
+        all_procs = list_all_procedures()
+        avail_procs = list_available_procedures()
+        print("Available procedures:")
+        for proc_name in sorted(avail_procs):
+            version = get_procedure(proc_name).get_version()
+            if version is None:
+                version = "???"
+            print(f"{proc_name} v{version}")
 
+        print()
+        print("Other supported procedures:")
+        print(" ".join(sorted(all_procs - avail_procs)))
+        print()
+
+    if args["version"] or args["all"]:
+        info_version()
+    if args["programs"] or args["all"]:
+        info_programs()
+    if args["procedures"] or args["all"]:
+        info_procedures()
+    if args["config"] or args["all"]:
+        print(">>> Configuration information")
+        print()
+        print(global_repr())
+
+
+def data_arg_helper(data_arg: str) -> 'ResultInput':
+    """
+    Converts the data argument of run and run-procedure commands to a ResultInput for compute
+
+    Parameters
+    ----------
+    data_arg: str
+        Either a data blob or file name or '-' for STDIN
+
+    Returns
+    -------
+    ResultInput
+        An input for compute.
+    """
+    if data_arg == "-":
+        return ResultInput.parse_raw(sys.stdin.read())
```

Comment:
This will work for now, but we will need to become more sophisticated as I believe `sys.stdin.read()` will always write out a str which will be interpreted as JSON.

Reply:
I agree. I think this is something of an upstream issue in QCElemental, since it implements the logic of detecting MsgPack vs JSON vs YAML maybe in the future. Currently, I think it uses the extension (datatype) for files (raw data). 

Given some STDIN, I don't know how to know if it's bytes or str. It might make sense to detect the input type in a more sophisticated manner. Does MsgPack have a header? How far do we have to read into a JSON vs YAML file to figure out what kind it is?

We could also just do something like:
```
data = None
for parser in [yaml_parser, msgpack_parser, json_parser]:
    try:
        data = parser(blob)
        break
    except ParserError:
        pass
else:
    throw ValueError("Cannot parse data")
```

Reply:
What about `--input-encoding=None` where `None` would try everything? Might be nice to have a `--output-encoding` as well.

We might need to port over some CLI testing tech from Fractal.

Reply:
This is getting large enough that I would vote to merge the current and make an issue to prevent this PR from getting too big.

Reply:
Okay, I will do the `info config` suggestion and then we can merge this.

Reply:
Created issues #135 and #136 for this.